src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -0,0 +1,726 @@
+using Microsoft.ML.Runtime.Data.Conversion;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+
+[assembly: LoadableClass(PartitionedFileLoader.Summary, typeof(PartitionedFileLoader), typeof(PartitionedFileLoader.Arguments), typeof(SignatureDataLoader),
+    PartitionedFileLoader.UserName, PartitionedFileLoader.LoadName, PartitionedFileLoader.ShortName)]
+
+[assembly: LoadableClass(PartitionedFileLoader.Summary, typeof(PartitionedFileLoader), null, typeof(SignatureLoadDataLoader),
+    PartitionedFileLoader.UserName, PartitionedFileLoader.LoadName, PartitionedFileLoader.ShortName)]
+
+namespace Microsoft.ML.Runtime.Data
+{
+    /// <summary>
+    /// Loads a set of directory partitioned files into an IDataView.
+    /// The directories of the file will treated as column data and the underlying files are loaded using the data loader.
+    /// The first file will be used as the basis for all follow-up file paths and schemas. Any files that don't match
+    /// the expected path or schema will be skipped.
+    /// </summary>
+    /// <example>
+    /// Sample directory structure:
+    ///
+    /// Data/
+    ///     Year=2017/
+    ///         Month=01/
+    ///             data1.parquet
+    ///             data1.parquet
+    ///         Month=02/
+    ///             data1.parquet
+    ///             data1.parquet
+    ///     Year=2018/
+    ///         Month=01/
+    ///             data1.parquet
+    ///             data1.parquet
+    /// </example>
+    public sealed class PartitionedFileLoader : IDataLoader
+    {
+        internal const string Summary = "Loads a horizontally partitioned file set.";
+        internal const string UserName = "Partitioned Loader";
+        public const string LoadName = "PartitionedLoader";
+        public const string ShortName = "Part";
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "PARTLOAD",
+                verWrittenCur: 0x00010001, // Initial
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoadName);
+        }
+
+        public class Arguments
+        {
+            [Argument(ArgumentType.Required, HelpText = "Base path to the directory of your partitioned files.", ShortName = "bp")]
+            public string BasePath;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Append a column with the file path.", ShortName = "path")]
+            public bool IncludePathColumn = false;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Path parser to extract column name/value pairs from the file path.", ShortName = "parser")]
+            public IPartitionedPathParserFactory PathParserFactory = new ParquetPartitionedPathParserFactory();
+
+            [Argument(ArgumentType.Multiple, HelpText = "The data loader.")]
+            public SubComponent<IDataLoader, SignatureDataLoader> Loader;
```

Comment:
>public SubComponent<IDataLoader, SignatureDataLoader> Loader; [](start = 12, length = 61)

Wat's your strategy in case if one file couldn't be read by loader. Right now it looks like you just throw exception. Maybe make sense to add strategy how to handle this situation by throwing exception, or just warn user by telling them which files can't be readed?

Reply:
Added exception handling.